sourcegraph/cody
# Edit: Process streamed insertions in queue and fix range expansion

## Description

closes https://github.com/sourcegraph/jetbrains/issues/1449

requires https://github.com/sourcegraph/jetbrains/pull/1509

This PR improves how we stream insertions into the document:
- Updates how we handle the `insertionPromise` queue, to make it more consistent and easier to understand
- Fixes a problem where we would update `task.selectionRange` after an edit was applied, but incorrectly calculate the `characterDelta`. 
- Adds agent tests for generating and inserting a streamed edit from a cursor position

## Test plan

Tested with agent tests

Manual testing:
1. In JetBrains, create an edit with no selection
2. Check that the incoming edit is streamed into the document, without any duplicate characters or lines that break the inserted code

### Before


https://github.com/sourcegraph/cody/assets/9516420/55d30e17-6109-48c8-ab49-560e1e206309


### After

https://github.com/sourcegraph/cody/assets/9516420/fe630452-98c3-46c7-9d41-5dcdf2ce735c
